src/infra/azure/cosmos/upload-document.ts

```diff
@@ -0,0 +1,114 @@
+import { pipe } from "fp-ts/function";
+
+import * as E from "fp-ts/Either";
+import * as TE from "fp-ts/TaskEither";
+
+import {
+  BaseModel,
+  CosmosdbModel,
+  CosmosResource,
+} from "@pagopa/io-functions-commons/dist/src/utils/cosmosdb_model";
+
+import * as t from "io-ts";
+import { Container } from "@azure/cosmos";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+
+import {
+  AddUploadDocument,
+  GetUploadDocument,
+  UploadDocument,
+  uploadDocumentNotFoundError,
+  UpsertUploadDocument,
+} from "../../../signature-request/upload-document";
+import { EntityNotFoundError } from "../../../error";
+import { container } from "./database";
+
+const containerId = "upload-documents";
+
+const NewUploadDocument = t.intersection([UploadDocument, BaseModel]);
+
+type NewUploadDocument = t.TypeOf<typeof NewUploadDocument>;
+
+const RetrievedUploadDocumnet = t.intersection([
+  UploadDocument,
+  CosmosResource,
+]);
+
+type RetrievedUploadDocumnet = t.TypeOf<typeof RetrievedUploadDocumnet>;
+
+class UploadDocumentModel extends CosmosdbModel<
+  UploadDocument,
+  NewUploadDocument,
+  RetrievedUploadDocumnet
+> {
+  constructor(container: Container) {
+    super(container, NewUploadDocument, RetrievedUploadDocumnet);
+  }
+}
+
+const uploadDocumentModel = pipe(
+  container(containerId),
+  E.map((c) => new UploadDocumentModel(c))
+);
+
+const uploadDocumentModelTE = TE.fromEither(uploadDocumentModel);
+
+export const addUploadDocument: AddUploadDocument = (request) =>
+  pipe(
+    NewUploadDocument.decode(request),
+    E.mapLeft(() => uploadDocumentNotFoundError),
```

Comment:
I'm skeptic with this error handling where you're swallowing the reason of the failure (decoding issues). That would be better to have something like:

```suggestion
    E.mapLeft((err) => new Error(readableReporter(errs)),
```

Reply:
I attempted to implement in the same manner as [elsewhere](https://github.com/pagopa/io-functions-sign/blob/555aff409ea60e8212a50eab44f1187e93fc8f28/src/infra/azure/cosmos/signature-request.ts#L59 )

Reply:
`ReadableReporter` takes a validator as input. For this we have the `validate` function of **handler-kit** but it is not usable in this case as we have an Error and not ValidationError.

Reply:
readableReporter takes  `t.ValidationError[]` as input which is exactly the left part in `decode()`. I've updated the suggestion.

Reply:
For now I have mitigated it using the `validate` function of `handler-kit`. 
I leave it unresolved for future reference!